lib/seahorn/Smt/TypeChecker.cc

```diff
@@ -6,41 +6,134 @@
 #include "llvm/Support/raw_ostream.h"
 
 using namespace expr;
-
 namespace {
 // local code in this namespace
-
 //==-- main implementation goes here --==//
 class TCVR {
 
+  ExprMap m_cache;
+  std::map<Expr, ExprSet> m_endExpr;
+
+  bool m_isWellFormed;
+  Expr m_errorExp;
+
+  TypeChecker *const m_tc;
+  Expr m_topMost;
+
+  void reset(Expr exp) {
+    if (m_isWellFormed)
+      m_errorExp = Expr();
+
+    m_topMost = Expr();
+    m_isWellFormed = true;
+    m_cache.clear();
+  }
+
+  bool isConst(Expr exp) {
+    return bind::isBoolConst(exp) || bind::isIntConst(exp) ||
+           bind::isRealConst(exp) || bind::isUnintConst(exp) ||
+           bind::isArrayConst(exp) || bind::isStructConst(exp) ||
+           bv::isBvConst(exp);
+  }
+
+  bool isValue(Expr exp) { return isOpX<TRUE>(exp) || isOpX<FALSE>(exp); }
```

Comment:
looks like a bunch of this is old code that needs to be cleaned up. I will ignore anything that I believe to be stale

Reply:
So I think I’d like your input on this if possible. I’m not sure if the way I’m going about this is good.

I used this as a part of finding the end of an expression (consts, vars, etc.) so that I can check if the bounded variables in quantifiers and functions match the variables in the body. One I reach a simple type (const, vars …) I put it in a map. This seems hard coded to me but I’m not really sure how else to approach this.

Reply:
I do not understand the problem that you are solving. Can you be more explicit?

Perhaps you need to know whether an expression is a value or an operator. If so, this is a property of the operator and should be moved there.

Reply:
So I guess I could just make a generic function to check if its , for example, a const but I'm not sure if there's other types that could be used as bounded variables in quantifiers that I'm forgetting about

Reply:
sorry I made that comment without refreshing the page.  I didn't realize that you had responded. if for example you had for all (a b (a && b &&c)), c is not bounded so that would be an error

Reply:
Ya I think checking whether it is a value or operator would work, I'll add that in

Reply:
I still don't understand what you are doing. The only bound variables that can appear under a binder are `BVar`, all other terminals are constants and cannot be bound.

Reply:
the format for a binder is `(binder (a:Int b:Int c:Int) body))`, the decls `a`, `b`, and `c` are just arbitrary names. They do not appear in `body`. Technically they can, but their meaning would be different. The only important part of them is the sort. Perhaps that is the source of confusion.

Reply:
Oh I see, I guess I didn't understand the structure of it. Ill fix that which should also fix the other comments about the expensive set copy and mapping

Reply:
Well, the code is not super documented (please help in your code!), so it is not easy for you to figure out on your own. Ask questions when in doubt.